File: library/src/main/java/lib/virgo/lib_scrollcenteritem/ScrollToCenterView.java
package lib.virgo.lib_scrollcenteritem;

//public class ScrollToCenterView extends RecyclerView {
//
//    private static final int DEFAULT_ORIENTATION = 0;
//
//    private ScrollToCenterLayoutManager layoutManager;
//
//    private ScrollStateChangeListener scrollStateChangeListener;
//    private OnItemChangedListener onItemChangedListener;
//
//    public ScrollToCenterView(Context context) {
//        super(context);
//        init(null);
//    }
//
//    public ScrollToCenterView(Context context, AttributeSet attrs) {
//        super(context, attrs);
//        init(attrs);
//    }
//
//    public ScrollToCenterView(Context context, AttributeSet attrs, int defStyleAttr) {
//        super(context, attrs, defStyleAttr);
//        init(attrs);
//    }
//
//    private void init(AttributeSet attrs) {
//        int orientation = DEFAULT_ORIENTATION;
//        if (attrs != null) {
//            TypedArray ta = getContext().obtainStyledAttributes(attrs, R.styleable.ScrollToCenterView);
//            orientation = ta.getInt(R.styleable.ScrollToCenterView_stc_orientation, DEFAULT_ORIENTATION);
//            ta.recycle();
//        }
//
//        OrientationView o = OrientationView.HORIZONTAL;
//        if (orientation == 1) o = OrientationView.VERTICAL;
//        layoutManager = new ScrollToCenterLayoutManager(
//                getContext(), new ScrollStateListener(),
//                o);
//        setLayoutManager(layoutManager);
//    }
//
//    @Override
//    public void setLayoutManager(LayoutManager layout) {
//        if (layout instanceof ScrollToCenterLayoutManager) {
//            super.setLayoutManager(layout);
//        } else {
//            throw new IllegalArgumentException(getContext().getString(R.string.dsv_ex_msg_dont_set_lm));
//        }
//    }
//
//
//    @Override
//    public boolean fling(int velocityX, int velocityY) {
//        boolean isFling = super.fling(velocityX, velocityY);
//        if (isFling) {
//            layoutManager.onFling(velocityX, velocityY);
//        } else {
//            layoutManager.returnToCurrentPosition();
//        }
//        return isFling;
//    }
//
//    @Nullable
//    public ViewHolder getViewHolder(int position) {
//        View view = layoutManager.findViewByPosition(position);
//        return view != null ? getChildViewHolder(view) : null;
//    }
//
//    /**
//     * @return adapter position of the current item or -1 if nothing is selected
//     */
//    public int getCurrentItem() {
//        return layoutManager.getCurrentPosition();
//    }
//
//    public void setItemTransformer(ScrollItemTransformer transformer) {
//        layoutManager.setItemTransformer(transformer);
//    }
//
//    public void setItemTransitionTimeMillis(@IntRange(from = 10) int millis) {
//        layoutManager.setTimeForItemSettle(millis);
//    }
//
//    public void setOrientation(OrientationView orientation) {
//        layoutManager.setOrientation(orientation);
//    }
//
//    public void setOffscreenItems(int items) {
//        layoutManager.setOffscreenItems(items);
//    }
//
//    public void setScrollStateChangeListener(ScrollStateChangeListener<?> scrollStateChangeListener) {
//        this.scrollStateChangeListener = scrollStateChangeListener;
//    }
//
//    public void setScrollListener(ScrollListener<?> scrollListener) {
//        setScrollStateChangeListener(new ScrollListenerAdapter(scrollListener));
//    }
//
//    public void setOnItemChangedListener(OnItemChangedListener<?> onItemChangedListener) {
//        this.onItemChangedListener = onItemChangedListener;
//    }
//
//    private class ScrollStateListener implements IScrollStateListener {
//
//        @Override
//        public void onIsBoundReachedFlagChange(boolean isBoundReached) {
//            setOverScrollMode(isBoundReached ? OVER_SCROLL_ALWAYS : OVER_SCROLL_NEVER);
//        }
//
//        @Override
//        public void onScrollStart() {
//            if (scrollStateChangeListener != null) {
//                int current = layoutManager.getCurrentPosition();
//                ViewHolder holder = getViewHolder(current);
//                if (holder != null) {
//                    scrollStateChangeListener.onScrollStart(holder, current);
//                }
//            }
//        }
//
//        @Override
//        public void onScrollEnd() {
//            ViewHolder holder = null;
//            int current = layoutManager.getCurrentPosition();
//            if (scrollStateChangeListener != null) {
//                holder = getViewHolder(current);
//                if (holder == null) {
//                    return;
//                }
//                scrollStateChangeListener.onScrollEnd(holder, current);
//            }
//            if (onItemChangedListener != null) {
//                if (holder == null) {
//                    holder = getViewHolder(current);
//                }
//                if (holder != null) {
//                    onItemChangedListener.onCurrentItemChanged(holder, current);
//                }
//            }
//        }
//
//        @Override
//        public void onScroll(float currentViewPosition) {
//            if (scrollStateChangeListener != null) {
//                int current = getCurrentItem();
//                ViewHolder currentHolder = getViewHolder(getCurrentItem());
//
//                int newCurrent = current + (currentViewPosition < 0 ? 1 : -1);
//                ViewHolder newCurrentHolder = getViewHolder(newCurrent);
//
//                if (currentHolder != null && newCurrentHolder != null) {
//                    scrollStateChangeListener.onScroll(
//                            currentViewPosition, currentHolder,
//                            newCurrentHolder);
//                }
//            }
//        }
//
//        @Override
//        public void onCurrentViewFirstLayout() {
//            if (onItemChangedListener != null) {
//                int current = layoutManager.getCurrentPosition();
//                ViewHolder currentHolder = getViewHolder(current);
//                if (currentHolder != null) {
//                    onItemChangedListener.onCurrentItemChanged(currentHolder, current);
//                }
//            }
//        }
//    }
//}

File: library/src/main/java/lib/virgo/lib_scrollcenteritem/ScrollToCenterView.kt
package lib.virgo.lib_scrollcenteritem

import android.content.Context
import android.util.AttributeSet
import android.view.View
import androidx.annotation.IntRange
import androidx.recyclerview.widget.RecyclerView
import lib.virgo.lib_scrollcenteritem.callback.OnItemChangedListener
import lib.virgo.lib_scrollcenteritem.callback.ScrollStateChangeListener
import androidx.core.content.withStyledAttributes
import lib.virgo.lib_scrollcenteritem.callback.IScrollStateListener
import lib.virgo.lib_scrollcenteritem.callback.ScrollListener
import lib.virgo.lib_scrollcenteritem.transform.ScrollItemTransformer
import lib.virgo.lib_scrollcenteritem.util.ScrollListenerAdapter

/**
 * Rebuilt from [DroidSky/DiscreteScrollView](https://github.com/DroidSky/DiscreteScrollView)
 */
class ScrollToCenterView @JvmOverloads constructor(
    private val context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0,
) : RecyclerView(context, attrs, defStyleAttr) {

    companion object {
        private const val DEFAULT_ORIENTATION = 0
    }

    private val layoutManager: ScrollToCenterLayoutManager
    private var scrollStateChangeListener: ScrollStateChangeListener<ViewHolder>? = null

    init {
        var orientation = DEFAULT_ORIENTATION

        context.withStyledAttributes(attrs, R.styleable.ScrollToCenterView) {
            orientation = getInt(R.styleable.ScrollToCenterView_stc_orientation, DEFAULT_ORIENTATION)
        }
        val o = if (orientation == 1) OrientationView.VERTICAL else OrientationView.HORIZONTAL
        layoutManager = ScrollToCenterLayoutManager(context, ScrollStateListener(), o)
        super.setLayoutManager(layoutManager)
    }

    override fun setLayoutManager(layout: LayoutManager?) {
        if (layout is ScrollToCenterLayoutManager) {
            super.setLayoutManager(layout)
        } else {
            throw IllegalArgumentException(
                context.getString(R.string.dsv_ex_msg_dont_set_lm)
            )
        }
    }

    override fun fling(velocityX: Int, velocityY: Int): Boolean {
        val isFling = super.fling(velocityX, velocityY)
        if (isFling) {
            layoutManager.onFling(velocityX, velocityY)
        } else {
            layoutManager.returnToCurrentPosition()
        }
        return isFling
    }

    fun getViewHolder(position: Int): ViewHolder? {
        val view: View? = layoutManager.findViewByPosition(position)
        return view?.let { getChildViewHolder(it) }
    }

    /**
     * @return adapter position of the current item or -1 if nothing is selected
     */
    fun getCurrentItem(): Int = layoutManager.currentPosition

    fun setItemTransformer(transformer: ScrollItemTransformer?) {
        layoutManager.setItemTransformer(transformer)
    }

    fun setItemTransitionTimeMillis(@IntRange(from = 10) millis: Int) {
        layoutManager.setTimeForItemSettle(millis)
    }

    fun setOrientation(orientation: OrientationView) {
        layoutManager.setOrientation(orientation)
    }

    fun setOffscreenItems(items: Int) {
        layoutManager.setOffscreenItems(items)
    }

    fun setScrollStateChangeListener(listener: ScrollStateChangeListener<ViewHolder>?) {
        scrollStateChangeListener = listener
    }

    fun setScrollListener(scrollListener: ScrollListener<ViewHolder>) {
        setScrollStateChangeListener(ScrollListenerAdapter(scrollListener))
    }
    private var onItemChangedListener: OnItemChangedListener<ViewHolder>? = null
    fun setOnItemChangedListener(listener: OnItemChangedListener<ViewHolder>?) {
        onItemChangedListener = listener
    }

    private inner class ScrollStateListener : IScrollStateListener {

        override fun onIsBoundReachedFlagChange(isBoundReached: Boolean) {
            overScrollMode = if (isBoundReached) OVER_SCROLL_ALWAYS else OVER_SCROLL_NEVER
        }

        override fun onScrollStart() {
            scrollStateChangeListener?.let { listener ->
                val current = layoutManager.currentPosition
                val holder = getViewHolder(current)
                if (holder != null) {
                    listener.onScrollStart(holder, current)
                }
            }
        }

        override fun onScrollEnd() {
            val current = layoutManager.currentPosition
            var holder = getViewHolder(current)

            scrollStateChangeListener?.let { listener ->
                if (holder == null) return
                listener.onScrollEnd(holder, current)
            }

            onItemChangedListener?.let { listener ->
                if (holder == null) {
                    holder = getViewHolder(current)
                }
                if (holder != null) {
                    listener.onCurrentItemChanged(holder, current)
                }
            }
        }

        override fun onScroll(currentViewPosition: Float) {
            scrollStateChangeListener?.let { listener ->
                val current = getCurrentItem()
                val currentHolder = getViewHolder(current)

                val newCurrent = current + if (currentViewPosition < 0) 1 else -1
                val newCurrentHolder = getViewHolder(newCurrent)

                if (currentHolder != null && newCurrentHolder != null) {
                    listener.onScroll(currentViewPosition, currentHolder, newCurrentHolder)
                }
            }
        }

        override fun onCurrentViewFirstLayout() {
            onItemChangedListener?.let { listener ->
                val current = layoutManager.currentPosition
                val holder = getViewHolder(current)
                if (holder != null) {
                    listener.onCurrentItemChanged(holder, current)
                }
            }
        }
    }
}